examples/test_application/src/main.rs

```diff
@@ -0,0 +1,451 @@
+};
+
+/// Messages from main thread to Participant worker threads.
+#[derive(Debug, Copy, Clone, Eq, Hash, PartialEq)]
+enum SubProtocol {
+    KeyGeneration,
+    AuxInfo,
+    Presign,
+    Sign,
+}
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct CommandLineArgs {
+    /// Number of participant worker threads to use.
+    #[arg(short, long)]
+    number_of_workers: usize,
+    /// Number of times to perform each sub-protocol of tss-ecdsa.
+    /// protocol_executions > 1 useful for computing meaningful average
+    /// execution times.
+    #[arg(short, long)]
+    protocol_executions: usize,
+}
+
+/// Outgoing message from worker threads to main thread.
+#[derive(Debug)]
+enum MessageFromWorker {
+    /// Worker made progress on sub-protocol, returning some messages.
+    FinishedRound(Vec<Message>),
+    /// Current sub-protocol ended.
+    SubProtocolEnded,
+}
+
+/// Maps [`ParticipantIdentifier`] to [`Sender`] channels for routing message to
+/// the correct participant.
+type WorkerChannels = HashMap<ParticipantIdentifier, Sender<Message>>;
+/// Channels for informing workers of the next [`SubProtocol`] to perform with
+/// the given SID [`Identifier`].
+type NextActionChannels = Vec<Sender<(SubProtocol, Identifier)>>;
+
+/// 1) Set up logging.
+/// 2) Create MPSC channels for communication between the workers and main
+/// thread. 3) Spawns N participant/worker threads ensuring the channels are
+/// "connected" properly. 4) Main thread executes the entire [`tss_ecdsa`]
+/// protocol to sign a message.
+fn main() -> anyhow::Result<()> {
+    let cli = CommandLineArgs::parse();
+    // Set up logging.
+    let filter = EnvFilter::from_default_env().add_directive("test_application=info".parse()?);
+    tracing_subscriber::fmt()
+        .with_env_filter(filter)
+        .without_time()
+        .compact()
+        .init();
+    let span = span!(Level::INFO, "main");
+    let _enter = span.entered();
+
+    let num_workers = cli.number_of_workers;
+    let rng = &mut OsRng;
+
+    // Channel from workers to main thread (workers us MPSC).
+    let (outgoing_tx, workers_rx) = channel::<MessageFromWorker>();
+    let mut worker_messages: WorkerChannels = HashMap::new();
+    let mut next_action: NextActionChannels = Vec::new();
+
+    // Assign a unique identifier to each participant to uniquely identify them
+    // through the sub-protocols.
+    let mut participants: Vec<ParticipantIdentifier> = Vec::new();
+    for _ in 0..num_workers {
+        participants.push(ParticipantIdentifier::random(rng));
+    }
+
+    info!("Spawning {num_workers} worker threads");
+
+    // Spawn worker threads. Link worker to main thread with channels.
+    for i in 0..num_workers {
+        // Create `ParticipantConfig`.
+        let id = participants[i];
+        let mut other_ids = participants.clone();
+        other_ids.remove(i);
+        let config = ParticipantConfig { id, other_ids };
```

Comment:
thought: should we make the `random_quorum` function on `ParticipantConfig` public?

Reply:
That makes sense! I added it to this PR and fixed the doc string.